src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +370,74 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
+        {
+            if (_loadedProjects.ContainsKey(projectPath))
+            {
+                return;
+            }
+
+            _loadedProjects.Add(projectPath, new ProjectLoadState.BeginLoadingWithPrimordial(primordialProjectId));
+            _projectsToLoadAndReload.AddWork(new ProjectToLoad(projectPath, ProjectGuid: null, ReportTelemetry: true));
+        }
+    }
+
+    protected async Task LoadProjectsAsync(ImmutableArray<(string ProjectPath, string? ProjectGuid)> projectPaths, CancellationToken cancellationToken)
+    {
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            foreach (var (path, guid) in projectPaths)
+            {
+                if (_loadedProjects.ContainsKey(path))
+                {
+                    continue;
+                }
+
+                _loadedProjects.Add(path, new ProjectLoadState.BeginLoading());
+                _projectsToLoadAndReload.AddWork(new ProjectToLoad(Path: path, ProjectGuid: guid, ReportTelemetry: true));
+            }
+
+            // TODO: since this method is used to implement 'OpenSolution' and 'OpenProjects',
+            // it seems like currently loaded projects which are not included in 'projectPaths' should be unloaded.
+        }
+
+        await _projectsToLoadAndReload.WaitUntilCurrentBatchCompletesAsync();
+    }
+
+    protected async ValueTask UnloadProjectAsync(string projectPath)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            if (_loadedProjects.Remove(projectPath, out var loadState))
+            {
+                if (loadState is ProjectLoadState.BeginLoading)
+                {
+                    // no work needed to unload this.
+                }
+                else if (loadState is ProjectLoadState.BeginLoadingWithPrimordial(var projectId))
+                {
+                    ProjectFactory.Workspace.OnProjectRemoved(projectId);
+                }
+                else if (loadState is ProjectLoadState.Loaded(var existingProjects))
+                {
+                    foreach (var existingProject in existingProjects)
+                    {
+                        existingProject.Dispose();
+                    }
+                }
+                else
+                {
+                    // TODO: we could also consider doing this in 'FileBasedProgramsProjectSystem.TryRemoveMiscellaneousDocument'.
+                    // Then we would delete the 'PrimordialOnly' type.
+                    // However, it seems nice to uniformly say: if a project is in 'Workspace', then it will be unloaded by this 'UnloadProjectAsync'.
+                    // Also, in the future, we may figure out how to do a design-time build for virtual files, which will make this a moot point.
+                    var primordialOnly = (ProjectLoadState.PrimordialOnly)loadState;
+                    ProjectFactory.Workspace.OnProjectRemoved(primordialOnly.PrimordialProjectId);
```

Comment:
See comment on the state definition earlier: this logic is the same as the other primordial state. We should really just collapse the states down since it's just creating additional risk of bugs here.

Reply:
(or if nothing else, share the logic for applying to the state)

Reply:
Done